internal/stats/metrics.go

```diff
@@ -1,29 +1,169 @@
 package stats
 
 import (
+	"fmt"
+
+	"github.com/edulinq/autograder/internal/config"
+	"github.com/edulinq/autograder/internal/log"
 	"github.com/edulinq/autograder/internal/timestamp"
 )
 
-type Metric interface {
+type MetricAttribute string
+type MetricType string
+
+// Keys for the attributes field inside of Metric and Query.
+const (
+	Unknown_Metric_Attribute_Key MetricAttribute = ""
+	Assignment_ID_Key                            = "assignment"
+	Analysis_Type_Key                            = "analysis-type"
+	Course_ID_Key                                = "course"
+	Endpoint_Key                                 = "endpoint"
+	Locator_Key                                  = "locator"
+	Sender_Key                                   = "sender"
+	Task_Type_Key                                = "task-type"
+	User_Email_Key                               = "user"
+)
+
+// Values for the type field inside of Metric and Query.
+const (
+	Unknown_Metric_Attribute_Type MetricType = ""
+	API_Request_Stats_Type                   = "api-request"
+	Code_Analysis_Time_Stats_Type            = "code-analysis-time"
+	Grading_Time_Stats_Type                  = "grading-time"
+	Task_Time_Stats_Type                     = "task-time"
+)
+
+const ATTRIBUTES_KEY = "attributes"
+
+type TimestampedMetric interface {
 	GetTimestamp() timestamp.Timestamp
 }
 
-type BaseMetric struct {
+type Metric struct {
 	Timestamp timestamp.Timestamp `json:"timestamp"`
 
+	Type MetricType `json:"type"`
+
+	Value float64 `json:"value"`
+
 	// Additional attributes that are not standard enough to be formalized in fields.
-	Attributes map[string]any `json:"attributes,omitempty"`
+	Attributes map[MetricAttribute]any `json:"attributes,omitempty"`
 }
 
-func (this BaseMetric) GetTimestamp() timestamp.Timestamp {
+func (this Metric) GetTimestamp() timestamp.Timestamp {
 	return this.Timestamp
 }
 
-type SystemMetrics struct {
-	BaseMetric
+func (this *Metric) addAttr(attrs *[]*log.Attr, key MetricAttribute) {
+	val, ok := this.Attributes[key]
+	if ok {
+		*attrs = append(*attrs, log.NewAttr(string(key), val))
+	}
+}
+
+func (this *Metric) LogValue() []*log.Attr {
+	attrs := []*log.Attr{
+		log.NewAttr("metric-type", this.Type),
+	}
+
+	this.addAttr(&attrs, Assignment_ID_Key)
+	this.addAttr(&attrs, Analysis_Type_Key)
+	this.addAttr(&attrs, Course_ID_Key)
+	this.addAttr(&attrs, Endpoint_Key)
+	this.addAttr(&attrs, Locator_Key)
+	this.addAttr(&attrs, Sender_Key)
+	this.addAttr(&attrs, Task_Type_Key)
+	this.addAttr(&attrs, User_Email_Key)
+
+	return attrs
+}
+
+func (this *Metric) Validate() error {
+	if this == nil {
+		return fmt.Errorf("No metric was given.")
+	}
+
+	if this.Type == Unknown_Metric_Attribute_Type {
+		return fmt.Errorf("Metric attribute was not set.")
+	}
```

Comment:
You should also verify that it is a known type.

With enums, it is not unusual to have a map around that can quickly be used to see if the value is known.
For example:
https://github.com/edulinq/autograder-server/blob/main/internal/model/role_server.go#L25
If you don't have another purpose for having the map (like the roles do),
then you can just have the value be `true`.